File: examples/csharp/avalonia/NotesShowcase/ViewModels/NoteFormVM.cs
using System.Reactive.Concurrency;
using System.Reactive.Linq;
using System.Reactive.Subjects;
using System.Windows.Input;
using VMx.Builders;
using VMx.Capabilities;
using VMx.Commands;
using VMx.Components;
using VMx.Forms;
using VMx.Messages;
using VMx.Services;
using VMx.Notifications;
using NotesShowcase.Models;

namespace NotesShowcase.ViewModels;

/// <summary>
/// VM for the note editor pane.
///
/// VMx-API adaptation (plan §3.a.9): the plan asked to subclass
/// <c>FormVM&lt;NoteModel&gt;</c>, but <see cref="FormVM{TM}"/> is sealed.
/// Composition instead: this VM owns a <see cref="FormVM{TM}"/> in strict mode
/// and exposes its <see cref="Draft"/>, <see cref="Snapshot"/>,
/// <see cref="IsDirty"/>, <see cref="ApproveCommand"/>, <see cref="DenyCommand"/>
/// surface directly, layered with <see cref="IsValid"/>, tag mutation commands,
/// and a "Saved" notification post-approve.
/// </summary>
public sealed class NoteFormVM : ComponentVMBase, IReconstructable
{
    private const string TitleRequired = "Title is required.";
    private readonly INoteRepository _repo;
    private readonly INotificationHub? _notificationHub;
    private FormVM<NoteModel>? _form;
    private NoteModel? _bound;
    private string _tagDraft = string.Empty;
    private bool _ownDisposed;
    private readonly Subject<System.Reactive.Unit> _canExecuteTrigger = new();
    private readonly Subject<NoteModel> _onSaved = new();
    private readonly IDispatcher _dispatcher;
    private IDisposable? _approvedSub;

    /// <inheritdoc/>
    public override ViewModelType Type => ViewModelType.Component;

    /// <summary>Public hub accessor.</summary>
    public new IMessageHub Hub => base.Hub;

    /// <summary>True once a note is bound (form constructed).</summary>
    public bool HasBoundNote => _form is not null;

    /// <summary>
    /// Live, editable note (the form's mutable model). Returns the original
    /// bound note before <see cref="BindTo"/> has been called.
    /// </summary>
    public NoteModel Draft
    {
        get => _form?.Model ?? _bound ?? Empty;
        set
        {
            if (_form is null) return;
            _form.SetModel(value);
            EmitDraftChanges();
        }
    }

    /// <summary>The form's snapshot (advances on each successful Approve).</summary>
    public NoteModel Snapshot => _form?.Snapshot ?? _bound ?? Empty;

    // Phase 5.a binding gap #1 (parity with Phase 5.b / 5.c): NoteModel is an
    // immutable record, so Avalonia widgets cannot two-way bind through
    // ``Draft.Title``/``Draft.Body``/``Draft.Starred`` directly. Expose mutable
    // scalar accessors that rebuild the draft via ``with`` expressions so
    // ``TextBox`` / ``CheckBox`` two-way bindings round-trip into the form and
    // ``IsDirty`` / ``IsValid`` / ``ApproveCommand.CanExecute`` flip reactively.

    /// <summary>Two-way bindable title (proxies <see cref="Draft"/>.Title).</summary>
    public string Title
    {
        get => Draft.Title;
        set
        {
            if (_form is null || string.Equals(Draft.Title, value, StringComparison.Ordinal)) return;
            Draft = Draft with { Title = value };
        }
    }

    /// <summary>Two-way bindable body (proxies <see cref="Draft"/>.Body).</summary>
    public string Body
    {
        get => Draft.Body;
        set
        {
            if (_form is null || string.Equals(Draft.Body, value, StringComparison.Ordinal)) return;
            Draft = Draft with { Body = value };
        }
    }

    /// <summary>Two-way bindable starred flag (proxies <see cref="Draft"/>.Starred).</summary>
    public bool Starred
    {
        get => Draft.Starred;
        set
        {
            if (_form is null || Draft.Starred == value) return;
            Draft = Draft with { Starred = value };
        }
    }

    /// <summary>Read-only view of the draft tag list (mutate via tag commands).</summary>
    public IReadOnlyList<string> Tags => Draft.Tags;

    /// <summary>
    /// Comma-joined tag list — bind UI text labels to this so the rendered
    /// string is "alpha, beta" instead of an enumerable repr. Mirrors Py
    /// <c>tags_text</c> (Round-3 Important C-I1) and TS <c>tagsText</c>.
    /// </summary>
    public string TagsText => string.Join(", ", Draft.Tags);

    /// <summary>True when the draft differs from the snapshot.</summary>
    public bool IsDirty => _form?.IsDirty ?? false;

    /// <summary>True when the draft passes FormVM validation.</summary>
    public bool IsValid => _form?.IsValid ?? false;

    /// <summary>Field-level validation error for <see cref="Title"/>, if any.</summary>
    public string? TitleError => _form?.FieldError(nameof(Title));

    /// <summary>Approve = persist via repo + publish "Saved" notification.</summary>
    public ICommand ApproveCommand { get; }

    /// <summary>Deny = revert draft to snapshot.</summary>
    /// <summary>
    /// Stable command that reverts the currently-bound form (no-op when
    /// unbound). One object for the VM's lifetime — and it re-emits this
    /// VM's own draft channels: the inner FormVM's Deny publishes with
    /// sender = FormVM, which the XAML bindings (keyed on this VM) never
    /// observe, so the editor kept the edited text on screen (real-wiring
    /// audit, pass 6).
    /// </summary>
    public ICommand DenyCommand { get; }

    /// <summary>
    /// Emits the persisted note after each successful save — the workspace
    /// refreshes the matching list row (Title/Starred were
    /// construction-time snapshots otherwise).
    /// </summary>
    public IObservable<NoteModel> OnSaved => _onSaved.AsObservable();

    /// <summary>Buffer text for "add tag" input.</summary>
    public string TagDraft
    {
        get => _tagDraft;
        set
        {
            if (string.Equals(_tagDraft, value, StringComparison.Ordinal)) return;
            _tagDraft = value;
            Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(TagDraft)));
            RaisePropertyChanged(nameof(TagDraft));
            _canExecuteTrigger.OnNext(System.Reactive.Unit.Default);
        }
    }

    /// <summary>Append <see cref="TagDraft"/> to the draft's tag list.</summary>
    public ICommand AddTagCommand { get; }

    /// <summary>Remove a tag from the draft's tag list.</summary>
    public ICommand RemoveTagCommand { get; }

    /// <summary>
    /// Binds the form to <paramref name="note"/> (creates / replaces the inner
    /// <see cref="FormVM{TM}"/> with a fresh snapshot).
    /// </summary>
    public void BindTo(NoteModel note)
    {
        _form?.Dispose();
        _approvedSub?.Dispose();
        _bound = note;
        _form = new FormVM<NoteModel>(
            initial: note,
            persister: PersistAsync,
            hub: Hub,
            strict: true,
            validators: new Dictionary<string, Func<NoteModel, string?>>
            {
                [nameof(Title)] = note => string.IsNullOrWhiteSpace(note.Title) ? TitleRequired : null
            });
        _approvedSub = _form.OnApproved.Subscribe(m => _onSaved.OnNext(m));
        EmitDraftChanges();
    }

    /// <summary>
    /// Clears the form back to its initial empty state — disposes the inner
    /// <see cref="FormVM{TM}"/>, resets <see cref="HasBoundNote"/> to
    /// <c>false</c>, and emits PropertyChanged for the bound surface so
    /// widgets re-read (Title / Body / Starred / Tags / TagsText all flip
    /// to the empty model). Round-4 Important-1: called by
    /// <see cref="WorkspaceVM"/> when <see cref="NotesViewVM.Current"/>
    /// transitions to <c>null</c> (e.g. the selected note is deleted) so
    /// the editor does not display ghost data from the just-removed note.
    ///
    /// Round-5 Minor: also reset <see cref="TagDraft"/>. The user-typed
    /// tag input buffer is part of the editor state, so a binding
    /// transition must clear it too — otherwise the chip input still shows
    /// the orphan text after the note disappears. Cross-flavor parity with
    /// Python <c>self._tag_draft = ""</c> and TS <c>this.tagDraft = ""</c>.
    /// </summary>
    public void Unbind()
    {
        var hadTagDraft = _tagDraft.Length > 0;
        if (_form is null && _bound is null && !hadTagDraft) return;
        _form?.Dispose();
        _form = null;
        _bound = null;
        if (hadTagDraft)
        {
            _tagDraft = string.Empty;
            Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(TagDraft)));
            RaisePropertyChanged(nameof(TagDraft));
        }
        EmitDraftChanges();
    }

    /// <summary>
    /// Awaitable approve cycle — persists via the repo and (on success)
    /// publishes a "Saved" notification. Useful in tests.
    /// </summary>
    public async Task ApproveAsync()
    {
        if (_form is null) return;
        await _form.ApproveAsync().ConfigureAwait(false);
        // This continuation runs off the UI thread (ConfigureAwait(false));
        // EmitDraftChanges raises INPC into live XAML bindings, so marshal
        // (real-wiring audit, pass 6).
        _dispatcher.Foreground.Schedule(() =>
        {
            if (_ownDisposed) return; // queued tail may outlive the VM
            EmitDraftChanges();
            if (_notificationHub is not null)
            {
                _ = _notificationHub.Post(new Notification(
                    NotificationType.Notification,
                    $"Saved “{Snapshot.Title}”"));
            }
        });
    }

    private async Task PersistAsync(NoteModel note)
    {
        await _repo.SaveNoteAsync(note).ConfigureAwait(false);
    }

    private void AddTag()
    {
        if (string.IsNullOrWhiteSpace(_tagDraft)) return;
        var trimmed = _tagDraft.Trim();
        if (Draft.Tags.Contains(trimmed, StringComparer.OrdinalIgnoreCase)) return;
        Draft = Draft with { Tags = Draft.Tags.Concat(new[] { trimmed }).ToArray() };
        TagDraft = string.Empty;
    }

    private void RemoveTag(string? tag)
    {
        if (string.IsNullOrEmpty(tag)) return;
        Draft = Draft with { Tags = Draft.Tags.Where(t => !string.Equals(t, tag, StringComparison.OrdinalIgnoreCase)).ToArray() };
    }

    private void EmitDraftChanges()
    {
        // Includes the per-field scalar accessors (Title/Body/Starred/Tags) so
        // two-way bound widgets receive PropertyChanged and re-read after every
        // draft mutation. See Phase 5.a binding gap #1.
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(Draft)));
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(Snapshot)));
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(IsDirty)));
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(IsValid)));
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(TitleError)));
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(Title)));
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(Body)));
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(Starred)));
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(Tags)));
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(TagsText)));
        // Round-3 Important B-I2: both commands are stable objects now, but
        // consumers that re-resolve on change notifications still expect the
        // signal on rebinds (cross-flavor parity).
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(ApproveCommand)));
        Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(DenyCommand)));
        RaisePropertyChanged(nameof(Draft));
        RaisePropertyChanged(nameof(Snapshot));
        RaisePropertyChanged(nameof(IsDirty));
        RaisePropertyChanged(nameof(IsValid));
        RaisePropertyChanged(nameof(TitleError));
        RaisePropertyChanged(nameof(Title));
        RaisePropertyChanged(nameof(Body));
        RaisePropertyChanged(nameof(Starred));
        RaisePropertyChanged(nameof(Tags));
        RaisePropertyChanged(nameof(TagsText));
        RaisePropertyChanged(nameof(ApproveCommand));
        RaisePropertyChanged(nameof(DenyCommand));
        _canExecuteTrigger.OnNext(System.Reactive.Unit.Default);
    }

    private NoteFormVM(
        string name,
        string hint,
        IMessageHub hub,
        IDispatcher dispatcher,
        INoteRepository repo,
        INotificationHub? notificationHub)
        : base(name, hint, hub, dispatcher, onConstruct: null, onDestruct: null)
    {
        _repo = repo;
        _notificationHub = notificationHub;
        _dispatcher = dispatcher;
        // Phase 5.a binding gap #1: draft mutation must flip
        // ApproveCommand.CanExecute reactively for Avalonia buttons. Wire the
        // ``_canExecuteTrigger`` subject through .Triggers(...) so each
        // EmitDraftChanges call fires CanExecuteChanged.
        ApproveCommand = RelayCommand.Builder()
            .Predicate(() => IsDirty && IsValid)
            .Task(() => _ = ApproveAsync())
            .Triggers(_canExecuteTrigger)
            .Build();
        DenyCommand = RelayCommand.Builder()
            .Task(() =>
            {
                _form?.DenyCommand.Execute(null);
                EmitDraftChanges();
            })
            .Build();
        AddTagCommand = RelayCommand.Builder()
            .Predicate(() => HasBoundNote && !string.IsNullOrWhiteSpace(_tagDraft))
            .Task(AddTag)
            .Triggers(_canExecuteTrigger)
            .Build();
        RemoveTagCommand = RelayCommand<string?>.Builder()
            .Predicate(t => HasBoundNote && !string.IsNullOrEmpty(t))
            .Task(RemoveTag)
            .Build();
    }

    private static readonly NoteModel Empty = new(
        Id: "",
        NotebookId: "",
        Title: "",
        Tags: Array.Empty<string>(),
        Body: "",
        Starred: false,
        CreatedAt: DateTimeOffset.MinValue,
        UpdatedAt: DateTimeOffset.MinValue);

    /// <inheritdoc/>
    public override void Dispose()
    {
        if (_ownDisposed) { base.Dispose(); return; }
        _ownDisposed = true;
        _form?.Dispose();
        _approvedSub?.Dispose();
        (ApproveCommand as IDisposable)?.Dispose();
        (DenyCommand as IDisposable)?.Dispose();
        (AddTagCommand as IDisposable)?.Dispose();
        (RemoveTagCommand as IDisposable)?.Dispose();
        _canExecuteTrigger.OnCompleted();
        _canExecuteTrigger.Dispose();
        _onSaved.OnCompleted();
        _onSaved.Dispose();
        base.Dispose();
    }

    /// <summary>Returns a new empty builder.</summary>
    public static NoteFormVMBuilder Builder() => NoteFormVMBuilder.Empty;

    /// <summary>Immutable fluent builder.</summary>
    public sealed class NoteFormVMBuilder
    {
        private readonly string? _name;
        private readonly string _hint;
        private readonly IMessageHub? _hub;
        private readonly IDispatcher? _dispatcher;
        private readonly INoteRepository? _repo;
        private readonly INotificationHub? _notificationHub;

        internal static readonly NoteFormVMBuilder Empty = new();
        private NoteFormVMBuilder() { _hint = ""; }
        private NoteFormVMBuilder(
            string? name, string hint,
            IMessageHub? hub, IDispatcher? dispatcher,
            INoteRepository? repo, INotificationHub? notificationHub)
        {
            _name = name; _hint = hint; _hub = hub; _dispatcher = dispatcher;
            _repo = repo; _notificationHub = notificationHub;
        }

        /// <summary>Sets the required Name.</summary>
        public NoteFormVMBuilder Name(string name) => new(name, _hint, _hub, _dispatcher, _repo, _notificationHub);
        /// <summary>Sets the optional Hint.</summary>
        public NoteFormVMBuilder Hint(string hint) => new(_name, hint, _hub, _dispatcher, _repo, _notificationHub);
        /// <summary>Sets the required Services.</summary>
        public NoteFormVMBuilder Services(IMessageHub hub, IDispatcher dispatcher) => new(_name, _hint, hub, dispatcher, _repo, _notificationHub);
        /// <summary>Sets the required Repository.</summary>
        public NoteFormVMBuilder Repository(INoteRepository repo) => new(_name, _hint, _hub, _dispatcher, repo, _notificationHub);
        /// <summary>Sets the optional NotificationHub (default: silent).</summary>
        public NoteFormVMBuilder NotificationHub(INotificationHub hub) => new(_name, _hint, _hub, _dispatcher, _repo, hub);

        /// <summary>Builds the VM after validation.</summary>
        public NoteFormVM Build()
        {
            BuilderValidationException.Require(_name, "Name");
            BuilderValidationException.Require(_hub, "Hub");
            BuilderValidationException.Require(_dispatcher, "Dispatcher");
            BuilderValidationException.Require(_repo, "Repository");
            return new NoteFormVM(_name!, _hint, _hub!, _dispatcher!, _repo!, _notificationHub);
        }
    }
}

File: examples/typescript/react/notes-showcase/src/viewmodels/noteFormVM.ts
/**
 * NoteFormVM — VM for the note editor pane.
 *
 * VMx-API adaptation (mirrors C# / Python flavors): VMx's `FormVM<TM>` is
 * suitable for composition but we layer extra validation, tag mutation
 * commands, and a post-approve notification on top. This VM owns a
 * `FormVM<NoteModel>` in strict mode (so `approveCommand.canExecute` already
 * gates on `isDirty`), and our own predicate additionally gates on `isValid`.
 *
 * `onApproved` persists via the repository (delegated through the FormVM's
 * `persister`) and then publishes a "Saved" notification.
 */
import { Subject, Subscription, type Observable } from "rxjs";

import {
  ComponentVMBase,
  declareCapabilities,
  DiscriminatorVM,
  FormVM,
  PropertyChangedMessage,
  RelayCommand,
  RelayCommandOf,
  SearchableState,
  ViewModelType,
  type ICommand,
  type ICommandOf,
  type IDispatcher,
  type IMessageHub,
} from "@thekaveh/vmx";
import {
  Notification,
  NotificationType,
  type INotificationHub,
} from "@thekaveh/vmx/notifications";

import type { NoteModel } from "../models/noteModel.js";
import type { INoteRepository } from "../models/noteRepository.js";

const SENTINEL = Symbol("not-set");
const TITLE_REQUIRED = "Title is required.";
export type EditorMode = "edit" | "preview";

const EMPTY: NoteModel = {
  id: "",
  notebookId: "",
  title: "",
  tags: [],
  body: "",
  starred: false,
  createdAt: new Date(0).toISOString(),
  updatedAt: new Date(0).toISOString(),
};

export class NoteFormVM extends ComponentVMBase {
  readonly #repo: INoteRepository;
  readonly #notificationHub: INotificationHub | null;
  readonly #approveCommand: RelayCommand;
  readonly #addTagCommand: RelayCommand;
  readonly #removeTagCommand: RelayCommandOf<string>;
  readonly #denyCommand: RelayCommand;
  readonly #showEditModeCommand: RelayCommand;
  readonly #showPreviewModeCommand: RelayCommand;
  readonly #onSaved = new Subject<NoteModel>();
  readonly #editorMode = new DiscriminatorVM<EditorMode>("edit");
  readonly #editorModeSub: Subscription;
  readonly #tagSearch: SearchableState<string>;
  readonly #tagSearchSub: Subscription;
  #form: FormVM<NoteModel> | null = null;
  #bound: NoteModel | null = null;
  #tagDraft = "";
  #tagCatalog: readonly string[] = [];
  #tagSuggestions: readonly string[] = [];

  constructor(opts: {
    name: string;
    hint: string;
    hub: IMessageHub;
    dispatcher: IDispatcher;
    repository: INoteRepository;
    notificationHub?: INotificationHub | null;
  }) {
    super({
      name: opts.name,
      hint: opts.hint,
      hub: opts.hub,
      dispatcher: opts.dispatcher,
    });
    this.#repo = opts.repository;
    this.#notificationHub = opts.notificationHub ?? null;
    declareCapabilities(this, "IReconstructable");
    this.#tagSearch = new SearchableState<string>({
      items: () => this.#tagCatalog,
      predicate: (tag, term) => {
        const normalized = term.trim().toLowerCase();
        if (normalized.length === 0) return false;
        if (!tag.toLowerCase().includes(normalized)) return false;
        return !this.draft.tags.some((existing) =>
          existing.toLowerCase() === tag.toLowerCase(),
        );
      },
      debounceMs: 0,
      scheduler: opts.dispatcher.foreground,
    });

    // Stable deny delegate (real-wiring audit, pass 6): the inner FormVM's
    // denyCommand publishes with sender = FormVM, which useVm's sender
    // filter drops — the DOM never re-rendered a revert. One stable command
    // delegates to the live form and re-emits this VM's own draft channels.
    this.#denyCommand = RelayCommand.builder()
      .task(() => {
        this.#form?.denyCommand.execute();
        this.#emitDraftChanges();
      })
      .build();

    this.#approveCommand = RelayCommand.builder()
      .predicate(() => this.isDirty && this.isValid)
      .task(() => {
        void this.approveAsync();
      })
      .build();
    this.#addTagCommand = RelayCommand.builder()
      .predicate(
        () => this.hasBoundNote && this.#tagDraft.trim().length > 0,
      )
      .task(() => this.#addTag())
      .build();
    this.#removeTagCommand = RelayCommandOf.builder<string>()
      .predicate((tag) => this.hasBoundNote && tag.length > 0)
      .task((tag) => this.#removeTag(tag))
      .build();
    this.#showEditModeCommand = RelayCommand.builder()
      .predicate(() => this.isPreviewMode)
      .task(() => this.#editorMode.setActiveKey("edit"))
      .triggers(this.#editorMode.activeChanged)
      .build();
    this.#showPreviewModeCommand = RelayCommand.builder()
      .predicate(() => this.isEditMode)
      .task(() => this.#editorMode.setActiveKey("preview"))
      .triggers(this.#editorMode.activeChanged)
      .build();
    this.#editorModeSub = this.#editorMode.activeChanged.subscribe(() => {
      this._hub.send(PropertyChangedMessage.create(this, this._name, "editorMode"));
      this._raisePropertyChanged("editorMode");
      this._hub.send(PropertyChangedMessage.create(this, this._name, "isPreviewMode"));
      this._raisePropertyChanged("isPreviewMode");
      this._hub.send(PropertyChangedMessage.create(this, this._name, "isEditMode"));
      this._raisePropertyChanged("isEditMode");
      this._hub.send(PropertyChangedMessage.create(this, this._name, "showEditModeCommand"));
      this._raisePropertyChanged("showEditModeCommand");
      this._hub.send(PropertyChangedMessage.create(this, this._name, "showPreviewModeCommand"));
      this._raisePropertyChanged("showPreviewModeCommand");
    });
    this.#tagSearchSub = this.#tagSearch.filtered.subscribe((suggestions) => {
      this.#tagSuggestions = suggestions;
      this.#emitTagSuggestionChanges();
    });
  }

  get type(): ViewModelType {
    return ViewModelType.Component;
  }

  get hub(): IMessageHub {
    return this._hub;
  }

  get hasBoundNote(): boolean {
    return this.#form !== null;
  }

  get editorMode(): EditorMode {
    return this.#editorMode.activeKey;
  }

  get isPreviewMode(): boolean {
    return this.#editorMode.isActive("preview");
  }

  get isEditMode(): boolean {
    return this.#editorMode.isActive("edit");
  }

  get showEditModeCommand(): ICommand {
    return this.#showEditModeCommand;
  }

  get showPreviewModeCommand(): ICommand {
    return this.#showPreviewModeCommand;
  }

  /** Live editable draft. Returns EMPTY before `bindTo`. */
  get draft(): NoteModel {
    return this.#form?.model ?? this.#bound ?? EMPTY;
  }

  set draft(value: NoteModel) {
    if (this.#form === null) return;
    this.#form.setModel(value);
    this.#emitDraftChanges();
  }

  /** Snapshot — advances on each successful approve. */
  get snapshot(): NoteModel {
    return this.#form?.snapshot ?? this.#bound ?? EMPTY;
  }

  get isDirty(): boolean {
    return this.#form?.isDirty ?? false;
  }

  get isValid(): boolean {
    return this.#form?.isValid ?? false;
  }

  get titleError(): string | null {
    return this.#form?.fieldError("title") ?? null;
  }

  /** Comma-joined tag list — bind UI text labels to this so the rendered
   * string is "alpha, beta" instead of an array repr. Mirrors Py
   * ``tags_text`` (Round-3 Important C-I1) and C# ``TagsText``. */
  get tagsText(): string {
    return this.draft.tags.join(", ");
  }

  get approveCommand(): ICommand {
    return this.#approveCommand;
  }

  get denyCommand(): ICommand {
    return this.#denyCommand;
  }

  /** Emits the persisted NoteModel after each successful save — the
   * workspace refreshes the matching list row (row labels / star /
   * filter inputs were construction-time snapshots otherwise). */
  get onSaved(): Observable<NoteModel> {
    return this.#onSaved.asObservable();
  }

  get tagDraft(): string {
    return this.#tagDraft;
  }

  set tagDraft(value: string) {
    if (this.#tagDraft === value) return;
    this.#tagDraft = value;
    this._hub.send(
      PropertyChangedMessage.create(this, this._name, "tagDraft"),
    );
    this._raisePropertyChanged("tagDraft");
    this.#tagSearch.searchTerm = value;
    this.#tagSearch.search();
  }

  get addTagCommand(): ICommand {
    return this.#addTagCommand;
  }

  get removeTagCommand(): ICommandOf<string> {
    return this.#removeTagCommand;
  }

  get tagSuggestions(): readonly string[] {
    return this.#tagSuggestions;
  }

  get tagSuggestionsText(): string {
    return this.#tagSuggestions.join(", ");
  }

  /** Binds the form to `note` (creates / replaces the inner FormVM). */
  bindTo(note: NoteModel): void {
    this.#form?.dispose();
    this.#bound = note;
    this.#form = new FormVM<NoteModel>({
      initial: note,
      persister: (m) => this.#persistAsync(m),
      hub: this._hub,
      strict: true,
      validators: {
        title: (m) => m.title.trim().length === 0 ? TITLE_REQUIRED : null,
      },
    });
    this.#form.onApproved.subscribe((m) => this.#onSaved.next(m));
    this.#emitDraftChanges();
  }

  /**
   * Clears the form back to its initial empty state — disposes the inner
   * FormVM, resets `hasBoundNote` to `false`, and emits PropertyChanged so
   * widgets re-read (draft / snapshot / tagsText flip to EMPTY).
   *
   * Round-4 Important-1: called by `WorkspaceVM` when `notesView.current`
   * transitions to `null` (e.g. the selected note is deleted in
   * `NotesViewVM.#deleteNoteAsync`) so the right-pane editor does not show
   * ghost data from the just-removed note. Mirrors C# `NoteFormVM.Unbind`
   * and Python `NoteFormVM.unbind`.
   *
   * Round-5 Minor: also reset ``tagDraft``. The user-typed tag input
   * buffer is part of the editor state, so a binding transition must
   * clear it too — otherwise the chip input still shows the orphan text
   * after the note disappears. Cross-flavor parity with C# `TagDraft =
   * string.Empty` and Python `self._tag_draft = ""`.
   */
  unbind(): void {
    const hadTagDraft = this.#tagDraft.length > 0;
    if (this.#form === null && this.#bound === null && !hadTagDraft) return;
    this.#form?.dispose();
    this.#form = null;
    this.#bound = null;
    if (hadTagDraft) {
      this.#tagDraft = "";
      this._hub.send(
        PropertyChangedMessage.create(this, this._name, "tagDraft"),
      );
      this._raisePropertyChanged("tagDraft");
    }
    this.#emitDraftChanges();
  }

  /**
   * Awaitable approve cycle — persists via the repo and (on success) publishes
   * a "Saved" notification. Useful in tests.
   */
  async approveAsync(): Promise<void> {
    if (this.#form === null) return;
    await this.#form.approveAsync();
    this.#emitDraftChanges();
    if (this.#notificationHub !== null) {
      void this.#notificationHub.post(
        new Notification(
          NotificationType.Notification,
          `Saved “${this.#form.snapshot.title}”`,
        ),
      );
    }
    await this.refreshTagSuggestionsAsync();
  }

  async refreshTagSuggestionsAsync(): Promise<void> {
    try {
      const snapshot = await this.#repo.loadAll();
      this.#tagCatalog = Array.from(
        new Set(
          snapshot.notes
            .flatMap((note) => note.tags)
            .map((tag) => tag.trim())
            .filter(Boolean),
        ),
      ).sort((left, right) => left.localeCompare(right));
      this.#tagSearch.search();
    } catch {
      this.#tagCatalog = [];
      this.#tagSearch.search();
    }
  }

  async #persistAsync(note: NoteModel): Promise<void> {
    await this.#repo.saveNote(note);
  }

  #addTag(): void {
    const trimmed = this.#tagDraft.trim();
    if (trimmed.length === 0) return;
    if (this.draft.tags.some((t) => t.toLowerCase() === trimmed.toLowerCase()))
      return;
    this.draft = { ...this.draft, tags: [...this.draft.tags, trimmed] };
    this.tagDraft = "";
    this.#tagSearch.search();
  }

  #removeTag(tag: string): void {
    if (tag.length === 0) return;
    this.draft = {
      ...this.draft,
      tags: this.draft.tags.filter(
        (t) => t.toLowerCase() !== tag.toLowerCase(),
      ),
    };
    this.#tagSearch.search();
  }

  #emitDraftChanges(): void {
    // Round-3 Important B-I2 parity: also fire for ``approveCommand`` /
    // ``denyCommand`` whose getters delegate to the inner ``#form`` (and to
    // ``#noopCommand`` before bindTo). Without this, bindings keep the
    // stale references after the form is rebound.
    //
    // Round-4 Minor-2 (cross-flavor parity): ``tagsText`` is a derived
    // accessor that re-projects on every draft mutation; without firing
    // PropertyChanged here any consumer subscribed specifically to
    // ``tagsText`` (e.g. a chip-strip label) would miss notifications.
    // Mirrors the C# emission list (``TagsText``) and the Python
    // ``tags_text`` DerivedProperty (which re-emits via ``_self_subject``).
    for (const name of [
      "draft",
      "snapshot",
      "isDirty",
      "isValid",
      "titleError",
      "tagsText",
      "tagSuggestions",
      "tagSuggestionsText",
      "approveCommand",
      "denyCommand",
    ]) {
      this._hub.send(
        PropertyChangedMessage.create(this, this._name, name),
      );
      this._raisePropertyChanged(name);
    }
  }

  #emitTagSuggestionChanges(): void {
    for (const name of ["tagSuggestions", "tagSuggestionsText"]) {
      this._hub.send(PropertyChangedMessage.create(this, this._name, name));
      this._raisePropertyChanged(name);
    }
  }

  protected override _onDispose(): void {
    this.#form?.dispose();
    this.#approveCommand.dispose();
    this.#addTagCommand.dispose();
    this.#removeTagCommand.dispose();
    this.#denyCommand.dispose();
    this.#showEditModeCommand.dispose();
    this.#showPreviewModeCommand.dispose();
    this.#editorModeSub.unsubscribe();
    this.#editorMode.dispose();
    this.#tagSearchSub.unsubscribe();
    this.#tagSearch.dispose();
    this.#onSaved.complete();
    super._onDispose();
  }

  static builder(): NoteFormVMBuilder {
    return new NoteFormVMBuilder();
  }
}

export class NoteFormVMBuilder {
  #name: string | null = null;
  #hint = "";
  #hub: IMessageHub | null = null;
  #dispatcher: IDispatcher | null = null;
  #repo: INoteRepository | typeof SENTINEL = SENTINEL;
  #notificationHub: INotificationHub | null = null;

  constructor(from?: NoteFormVMBuilder) {
    if (from) {
      this.#name = from.#name;
      this.#hint = from.#hint;
      this.#hub = from.#hub;
      this.#dispatcher = from.#dispatcher;
      this.#repo = from.#repo;
      this.#notificationHub = from.#notificationHub;
    }
  }

  name(value: string): NoteFormVMBuilder {
    const b = new NoteFormVMBuilder(this);
    b.#name = value;
    return b;
  }

  hint(value: string): NoteFormVMBuilder {
    const b = new NoteFormVMBuilder(this);
    b.#hint = value;
    return b;
  }

  services(hub: IMessageHub, dispatcher: IDispatcher): NoteFormVMBuilder {
    const b = new NoteFormVMBuilder(this);
    b.#hub = hub;
    b.#dispatcher = dispatcher;
    return b;
  }

  repository(repo: INoteRepository): NoteFormVMBuilder {
    const b = new NoteFormVMBuilder(this);
    b.#repo = repo;
    return b;
  }

  notificationHub(hub: INotificationHub): NoteFormVMBuilder {
    const b = new NoteFormVMBuilder(this);
    b.#notificationHub = hub;
    return b;
  }

  build(): NoteFormVM {
    if (this.#name === null) throw new Error("name is required");
    if (this.#hub === null || this.#dispatcher === null)
      throw new Error("services (hub + dispatcher) are required");
    if (this.#repo === SENTINEL) throw new Error("repository is required");
    return new NoteFormVM({
      name: this.#name,
      hint: this.#hint,
      hub: this.#hub,
      dispatcher: this.#dispatcher,
      repository: this.#repo,
      notificationHub: this.#notificationHub,
    });
  }
}
